Common/Thread/ParallelLoop.cpp

```diff
@@ -0,0 +1,142 @@
+		// Remember that stragglers are done on the current thread
+		// so we don't round up.
+		numTasks = (int)(totalFrac / delta);
+		printf("numTasks: %d\n", numTasks);
+
+		WaitableCounter *waitableCounter = new WaitableCounter(numTasks);
+		int64_t counter = (int64_t)lower << fractionalBits;
+
+		// Split up tasks as equitable as possible.
+		for (int i = 0; i < numTasks; i++) {
+			int start = (int)(counter >> fractionalBits);
+			int end = (int)((counter + delta) >> fractionalBits);
+			if (end > upper) {
+				// Let's do the stragglers on the current thread.
+				break;
+			}
+			threadMan->EnqueueTaskOnThread(i, new LoopRangeTask(waitableCounter, loop, start, end), TaskType::CPU_COMPUTE);
+			counter += delta;
+			if ((counter >> fractionalBits) > upper) {
+				break;
+			}
+		}
+
+		// Run stragglers on the calling thread directly.
+		// We might add a flag later to avoid this for some cases.
+		int stragglerStart = (int)(counter >> fractionalBits);
+		int stragglerEnd = upper;
+		if (stragglerStart < stragglerEnd) {
+			// printf("doing stragglers: %d-%d\n", start, upper);
+			loop(stragglerStart, stragglerEnd);
+		}
+		return waitableCounter;
+	}
+}
+
+void ParallelRangeLoop(ThreadManager *threadMan, const std::function<void(int, int)> &loop, int lower, int upper, int minSize) {
+	if (cpu_info.num_cores == 1 || (minSize >= (upper - lower) && upper > lower)) {
+		// "Optimization" for single-core devices, or minSize larger than the range.
+		// No point in adding threading overhead, let's just do it inline (since this is the blocking variant).
+		loop(lower, upper);
+		return;
+	}
+
+	if (minSize < 1) {
+		// There's no obvious value to default to.
+		minSize = 1;
+	}
+
+	WaitableCounter *counter = ParallelRangeLoopWaitable(threadMan, loop, lower, upper, minSize);
+	// TODO: Optimize using minSize. We'll just compute whether there's a remainer, remove it from the call to ParallelRangeLoopWaitable,
+	// and process the remainder right here. If there's no remainer, we'll steal a whole chunk.
+	if (counter) {
+		counter->WaitAndRelease();
+	}
+}
+
+// NOTE: Supports a max of 2GB.
+void ParallelMemcpy(ThreadManager *threadMan, void *dst, const void *src, size_t bytes) {
+	// This threshold can probably be a lot bigger.
+	if (bytes < 512) {
+		memcpy(dst, src, bytes);
+		return;
+	}
+
+	// 128 is the largest cacheline size on common CPUs.
+	// Still I suspect that the optimal minSize is a lot higher.
+
+	char *d = (char *)dst;
+	char *s = (char *)src;
+	ParallelRangeLoop(threadMan, [&](int l, int h) {
+		memmove(d + l, s + l, h - l);
+	}, 0, (int)bytes, 128);
+}
+
+// NOTE: Supports a max of 2GB.
+void ParallelMemset(ThreadManager *threadMan, void *dst, uint8_t value, size_t bytes) {
+	// This threshold can probably be a lot bigger.
+	if (bytes < 512) {
+		memset(dst, 0, bytes);
+		return;
+	}
+
+	// 128 is the largest cacheline size on common CPUs.
+	// Still I suspect that the optimal minSize is a lot higher.
```

Comment:
This one I didn't test at all, and just reused the minSize from memcpy, but I'd suspect it'd also perform better with larger values on all CPUs we target.

-[Unknown]